src/pkg/artifact/manager.go

```diff
@@ -30,44 +32,117 @@ type Manager interface {
 	// List artifacts according to the query, returns all artifacts if query is nil
 	List(ctx context.Context, query *q.Query) (total int64, artifacts []*Artifact, err error)
 	// Get the artifact specified by the ID
-	Get(ctx context.Context, id int64) (*Artifact, error)
+	Get(ctx context.Context, id int64) (artifact *Artifact, err error)
 	// Create the artifact. If the artifact is an index, make sure all the artifacts it references
 	// already exist
 	Create(ctx context.Context, artifact *Artifact) (id int64, err error)
 	// Delete just deletes the artifact record. The underlying data of registry will be
 	// removed during garbage collection
-	Delete(ctx context.Context, id int64) error
+	Delete(ctx context.Context, id int64) (err error)
 	// UpdatePullTime updates the pull time of the artifact
-	UpdatePullTime(ctx context.Context, artifactID int64, time time.Time) error
+	UpdatePullTime(ctx context.Context, artifactID int64, time time.Time) (err error)
 }
 
 // NewManager returns an instance of the default manager
 func NewManager() Manager {
-	return &manager{}
+	return &manager{
+		dao.New(),
+	}
 }
 
 var _ Manager = &manager{}
 
 type manager struct {
+	dao dao.DAO
 }
 
-func (m *manager) List(ctx context.Context, query *q.Query) (total int64, artifacts []*Artifact, err error) {
-	// TODO implement
-	return 0, nil, nil
+func (m *manager) List(ctx context.Context, query *q.Query) (int64, []*Artifact, error) {
+	total, err := m.dao.Count(ctx, query)
```

Comment:
the call to count can be removed, just use len(arts) to judge count.

Reply:
We can't, the `arts` only contains the elements which meet the query conditions

Reply:
correct, and that's why the `Count` should NEVER handles pagination and the flag `pagination` should be removed.